job_executor/worker/manager_state.py

```diff
@@ -0,0 +1,82 @@
+from job_executor.model.worker import Worker
+
+from typing import List
+
+
+REDUCED_WORKER_NUMBER = 2
+
+
+class ManagerState:
+    def __init__(self, default_max_workers=4, max_gb_all_workers=50):
+        """
+        :param default_max_workers: The maximum number of workers
+        :param max_gb_all_workers: Threshold in GB (50) for when the number
+        of workers are reduced
+        """
+        self.default_max_workers = default_max_workers
+        self.current_max_workers = default_max_workers
+        self.max_bytes_all_workers = (
+            max_gb_all_workers * 1024**3  # Threshold in bytes
+        )
+
+        self.workers: List[Worker] = []
+
+    @property
+    def dead_workers(self) -> List[Worker]:
+        """
+        Return a list of dead workers
+        """
+        return [worker for worker in self.workers if not worker.is_alive()]
+
+    @property
+    def alive_workers(self) -> List[Worker]:
+        """
+        Return a list of alive workers
+        """
+        return [worker for worker in self.workers if worker.is_alive()]
+
+    @property
+    def current_total_size(self) -> int:
+        return sum(
+            worker.job_size for worker in self.workers if worker.is_alive()
+        )
+
+    def can_spawn_new_worker(self, new_job_size: int) -> bool:
+        """
+        Called to check if a new worker can be spawned.
+        """
+        can_spawn = True
+
+        active_workers = len(self.alive_workers)
+        if active_workers >= self.current_max_workers:
+            can_spawn = False
+
+        return can_spawn
```

Comment:
```suggestion
    def can_spawn_new_worker(self, new_job_size: int) -> bool:
        """
        Called to check if a new worker can be spawned.
        """
        can_spawn = True

        active_workers = len(self.alive_workers)
        if active_workers >= self.current_max_workers:
            can_spawn = False
        if self.current_total_size + new_job_size >= self.max_bytes_all_workers:
            can_spawn = False
        return can_spawn
```
We now have the chance to check both these at the same time.
If we did something like this we wouldn't need the update_worker_limit, or a current/default max workers(just a max_workers) and it would never go past the MAX. (And it is arguably easier to read)

Reply:
🤔 ..yes this actually could simply much of the logic. Good idea!

Reply:
With this solution, it looks like big jobs will not be spawned until all the smaller jobs are done. This is not necessarily bad, but affects the import order and maybe user's expectations?

Reply:
That is true 👍🏻 If we want a system that actually stops jobs from breaking the limit we can either:
- Experience that bigger jobs can get "blocked" by smaller jobs as they can't fit the remaining space (current)
- Wait with including more jobs until the next job chronologically can fit the memory space

My 2 cents:
I think the first alternative (current) is less complicated and a more efficient use of time in most cases. If users complain that their jobs are getting picked up out of order we can consider implementing something more sophisticated at that point.

Reply:
Sounds good 👍 Let's wait and see